Formula/haste-client.rb

```diff
@@ -26,20 +26,29 @@ class HasteClient < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "f58ba7733cf6b73c7ad7be755d4d31c88af9577a43f9cd4658d8732bdd8bad3b"
   end
 
-  depends_on "ruby" if MacOS.version <= :sierra
+  on_linux do
+    depends_on "ruby"
 
-  resource "faraday" do
-    url "https://rubygems.org/gems/faraday-0.17.4.gem"
-    sha256 "11677b5b261fbbfd4d959f702078d81c0bb66006c00ab2f329f32784778e4d9c"
+    resource "json" do
+      url "https://rubygems.org/gems/json-2.5.1.gem"
+      sha256 "918d8c41dacb7cfdbe0c7bbd6014a5372f0cf1c454ca150e9f4010fe80cc3153"
+    end
   end
 
-  if MacOS.version <= :sierra
+  on_sierra :or_older do
```

Comment:
```suggestion
  on_os :linux, macos: :sierra_or_older do
```

or something would work for me. thoughts @carlocab @Rylan12?

Reply:
Fine with this! I prefer `on_system`, so that we can do `on_system :arm do`, and possibly a way to handle `&&`/`||` conditions, e.g. `on_system all_of: [:arm, macos: monterey_or_newer] do`.

Reply:
I'd rather we scope this tightly to what we actually need for now rather than making it more flexible than it needs to be. As far as I can tell from this PR, the only cases we need this for are "some macOS versions conditional on version, always on Linux" and nothing related to the macOS architecture, Linux version, etc.

Reply:
One of these would be easier to implement. Do any of them work? I'd like to avoid needing to completely re-design these again

```ruby
# Not sure which one makes more sense
on_linux_and_sierra :or_older
on_linux_or_sierra :or_older
```

```ruby
# Not sure which one makes more sense
on_sierra :or_older, :and_linux
on_sierra :or_older, :or_linux
```

I'd rather not but we could also just make `on_sierra :or_older` return true on Linux like we used to

Reply:
@Rylan12 I think we need a new name for the split OS case e.g. `on_system` or `on_os` or similar. Both `on_linux_and_sierra` and `on_sierra` are much harder to quickly parse.

Reply:
I'd also point out that there are a very limited number of situations where this combination OS case would be used. As far as I know, it only affects 8 formulae in any way. Of those, only 5 of the duplications are exclusively `depends_on` lines. 2 have a single duplicated resource, and 1 has a duplicated `fails_with` line. Are we really sure it's worth it for such a small number of formulae? And if so, can we improve it in other ways?

We may be able to modify the `depends_on`and `fails_with` DSLs to mimic `uses_from_macos "ruby", since: :high_sierra` so that we can specify which OS/Arches each dependency applies to. This might help reduce the duplicates and (maybe) make it easier to read.

Also, what about doing something like this for the two duplicated resources? It means you would still only have to bump the resource in one place. 

```ruby
resource do
  resource_url = "https://example.com"
  resource_sha256 = "..."

  on_linux do
    url resource_url
    sha256 resource_sha256
  end

  on_sierra :or_older do
    url resource_url
    sha256 resource_sha256
  end
end
```

Reply:
> Are you suggesting on_macos and on_linux stay, we add on_arm and on_intel, but use on_system or whatever it is for macOS versions?

Yup, `on_system` would just be a wrapper for cases where you want to do `on_macos` and `on_linux` at once.

> Or switch back to using the on_system method for everything?

Not this.

> fails_with DSLs to mimic

Personally: I think this DSL is kinda terrible so would rather not mimic it.

> Also, what about doing something like this for the two duplicated resources? It means you would still only have to bump the resource in one place.

If we need this only for resources, @carlocab is 🆗 with it and (eventually) `brew bump-formula-pr` etc. knows how to handle it: fine with me?

Reply:
Will discuss more in https://github.com/Homebrew/brew/pull/13529